tabs/newform.1
.\"
.\" SPDX-Licence-Identifier: CDDL-1.0
.\"  Copyright 1989 AT&T  Copyright (c) 1997, Sun Microsystems, Inc.  All Rights Reserved
.\" Portions Copyright (c) 2007 Gunnar Ritter, Freiburg i. Br., Germany
.\"
.\" Sccsid @(#)newform.1	1.8 (gritter) 2/3/07
.\" from OpenSolaris newform 1 "21 Jul 1997" "SunOS 5.11" "User Commands"
.TH NEWFORM 1 "2/3/07" "Heirloom Toolchest" "User Commands"
.SH NAME
newform \- change the format of a text file
.SH SYNOPSIS
.HP
.ad l
.nh
\fBnewform\fR [\fB\-s\fR]
[\fB\-i\fR\fItabspec\fR] [\fB\-o\fR\fItabspec\fR]
[\fB\-b\fR\fIn\fR] [\fB\-e\fR\fIn\fR] [\fB\-p\fR\fIn\fR]
[\fB\-a\fR\fIn\fR] [\fB\-f\fR] [\fB\-c\fR\fIchar\fR]
[\fB\-l\fR\fIn\fR] [\fIfilename\fR]...
.br
.ad b
.hy 1
.SH DESCRIPTION
\fBnewform\fR reads lines from the named \fIfilename\fRs, or the standard input if no input file is named, and reproduces the lines on the standard output.
Lines
are reformatted in accordance with command line options in effect.
.PP
Except for \fB\-s\fR, command line options may appear in any order, may be repeated, and may be intermingled with the optional \fIfilename\fRs.
Command line options are
processed in the order specified.
This means that option sequences like ``\fB\-e\fR15 \fB\-l\fR60'' will yield results different from ``\fB\-l\fR60 \fB\-e\fR15''.
Options
are applied to all \fIfilename\fRs on the command line.
.PP
The following options are supported:
.TP 10
\fB\-s\fR \fR
Shears off leading characters on each line up to the first tab and places up to 8 of the sheared
characters at the end of the line.
If more than 8 characters (not counting the first tab) are sheared, the eighth character is replaced by a \fB*\fR and any characters to the right of it are
discarded.
The first tab is always discarded.
.IP
An error message and program exit will occur if this option is used on a file without a tab on each line.
The characters sheared off are saved internally until all other options specified are applied
to that line.
The characters are then added at the end of the processed line.
.IP
For example, to convert a file with leading digits, one or more tabs, and text on each line, to a file beginning with the text, all tabs after the first expanded to spaces, padded with spaces out
to column 72 (or truncated to column 72), and the leading digits placed starting at column 73, the command would be:
.RS 14
\fBnewform \|\fR\fB\-s\fR\fB \|\fR\fB\-i\fR\fB \|\fR\fB\-l\fR\fB \|\fR\fB\-a\fR\fB \|\fR\fB\-e\fR\fB \|\fIfilename\fR
.RE
.TP 10
\fB\-i\fR\fItabspec\fR \fR
Input tab specification: expands tabs to spaces, according to the tab specifications given.
\fITabspec\fR recognizes all tab specification forms described in
.IR tabs (1).
In addition, \fItabspec\fR may be \fB\-\fR, in which \fBnewform\fR assumes that the tab specification is to be found in the first line read from the standard input (see
.IR fspec (5)).
If no \fItabspec\fR is given, \fItabspec\fR defaults to \fB\(mi8\fR\&.
A \fItabspec\fR of \fB\(mi0\fR expects no tabs; if any are found, they are treated as \fB\(mi1\fR\&.
.TP
\fB\-o\fR\fItabspec\fR \fR
Output tab specification: replaces spaces by tabs, according to the tab specifications given.
The tab specifications
are the same as for \fB\-i\fR\fItabspec\fR.
If no \fItabspec\fR is given, \fItabspec\fR defaults to \fB\(mi8\fR\&.
A \fItabspec\fR of \fB\(mi0\fR means that no spaces will be converted to tabs on output.
.TP
\fB\-b\fR\fIn\fR \fR
Truncate \fIn\fR characters from the beginning of the line when the line length is greater
than the effective line length (see \fB\(mil\fIn\fR).
Default is to truncate the number of characters necessary to obtain the effective line length.
The default value
is used when \fB\-b\fR with no \fIn\fR is used.
This option can be used to delete the sequence numbers from a \fBCOBOL\fR program as follows:
.RS 14
\fBnewform \|\fR\fB\-l\fR\fB1 \|\fR\fB\-b\fR\fB7 \|\fIfilename\fR
.RE
.TP 10
\fB\-e\fR\fIn\fR
Same as \fB\-b\fR\fIn\fR except that characters are truncated from the end
of the line.
.TP
\fB\-p\fR\fIn\fR
Prefix \fIn\fR characters (see \fB\-c\fR\fIchar\fR) to the beginning
of a line when the line length is less than the effective line length.
Default is to prefix the number of characters necessary to obtain the effective line length.
.TP
\fB\-a\fR\fIn\fR
Same as \fB\-p\fR\fIn\fR except characters are appended to the end of a line.
.TP
\fB\-f\fR
Write the tab specification format line on the standard output before any other lines are output.
The tab specification format line which is
printed will correspond to the format specified in the \fBlast\fR \fB\-o\fR option.
If no \fB\-o\fR option is specified, the line which is printed will contain the default
specification of \fB\(mi8\fR\&.
.TP
\fB\-c\fR\fIchar\fR
Change the prefix/append character to \fIchar\fR.
Default character for \fIchar\fR is a space.
.TP
\fB\-l\fR\fIn\fR
Set the effective line length to \fIn\fR characters.
If \fIn\fR is not
entered, \fB\-l\fR defaults to 72.
The default line length without the \fB\-l\fR option is 80 characters.
Note:
Tabs and backspaces are considered to be one character (use \fB\-i\fR
to expand tabs to spaces).
.IP
The \fB\(mil1\fR must be used to set the effective line length shorter than any existing line in the file so that the \fB\-b\fR option is activated.
.SH DIAGNOSTICS
All diagnostics are fatal.
.PD 0
.TP 29
\fBusage:\fR \fB \|.\|.\|.\fR \fR
\fBnewform\fR was called
with a bad option.
.TP
\fB"not \fR\fB\-s\fR\fB format"\fR
There was no tab on one line.
.TP
\fB"can't open file"\fR
Self-explanatory.
.TP
\fB"tabspec in error" \fR
A tab specification is incorrectly formatted, or specified tab stops are not ascending.
.TP
\fB"tabspec indirection illegal" \fR
A \fItabspec\fR read from a file (or standard input) may not contain a \fItabspec\fR referencing another file (or standard input).
.PD
.SH "SEE ALSO"
.IR csplit (1),
.IR tabs (1),
.IR fspec (5)
.SH NOTES
\fBnewform\fR normally only keeps track of physical characters; however, for the \fB\-i\fR and \fB\-o\fR options, \fBnewform\fR will keep track of backspaces
in order to line up tabs in the appropriate logical columns.
.PP
\fBnewform\fR will not prompt the user if a \fItabspec\fR is to be read from the standard input (by use of \fB\fR\fB\-i\fR\fB\-\fR or \fB\fR\fB\-o\fR\fB\-\fR).
.PP
If the \fB\-f\fR option is used, and the last \fB\-o\fR option specified was \fB\fR\fB\-o\fR\fB\-\fR, and was preceded by either a \fB\fR\fB\-o\fR\fB\-\fR or a \fB\fR\fB\-i\fR\fB\-\fR, the tab specification format line will be incorrect.


CHANGES
Release 250220
* seq, write: Fix memory fault due to illegal pointer access.
* cpio, pax: Fix manual page formatting.

Release 250209
* cpio: Fix manual page formatting.
* apply: Fixed the command string magic character computation algorithm that
  got garbled after a refactor made in the name of performance.
* build/mk.config, chroot, nawk: Adding -ltinfo for static compilation with
  glibc and making changes to adhere to C standards (patch by Gabriel Souza).
* [NEW] wall, write: New implementation ported from UNIX 7th (a.k.a. 32V),
  bulletproofed against buffer overflows and the CVE-2024-28085, which is
  available as the 'ssafe()' function at libcommon.
* rm: Alias '-e' flag to more known '-v' for verbose.
* date, shl: Fix compilation on platforms where (struct utmpx).ut_tv is not
  of type 'timeval'.
* rev: UTF-8 support, borrowed from suckless sbase. In varietate concordia.
* [NEW] apply, pick: Complete implementation for both apply and pick, from
  Research UNIX 8th and 10th respectively; extended payloads for pick in
  comparison with the original implementation.
* seq: Support for counting from negative numbers to positive or making a
  negative sequence.
* rev: Full refactor (made by Samuel Brederodes).
* Fixes for building dynamically on Arch Linux and other GNU C
  Library-based Linux systems which ships exclusively dynamic versions
  of its libraries, also remove `readlink_ucb' object file at `readlink/'
  build directory when executing `make clean' (patch by Molly A. McCollum).
* Remove `register's from legacy code that was living unchanged since at
  least the 80's, since modern linters started complaining about it and
  it is not used as much on most common architectures such as x86
  (patch by Samuel Brederodes).
* watch: `use_default_colors()', do not force black and white.
* [NEW] rev: New implementation, not capable of UTF-8 yet.
* cpio, diff, tabs: Fix duplicated definitions (made/reported by crumbtoo,
  merged by me, Luiz Antônio Rangel). Thanks! --- or better: Merci!
* rm: Added real support for the POSIX 1003.1(2008) standard, implemented
  IBM AIX(R)-style verbosing and 4.4BSD's '-d' option for directories ---
  rmdir(1)-esque payload --- as extensions to the standard.
* seq: Made its payload similar to UNIX v8's implementation, but with some
  extensions such as a fast algorithm for -w or -s --- which was already
  present before the (almost) full rewrite.
* timeout: avoid buffer overrun, also did some changes at the source code,
  such as cleaning up comments and initializing 'commandv' and 'fst_commandv'.
* watch: Used esr's deheader for removing unnecessary headers, also use
  pfmt(stderr, MM_ERROR, ...)  instead of prerror().
* watch, timeout: Some small corrections at the manual page.
* cpio: Skip over zip extra headers with fewer than four bytes (the minimum
  length for them to contain any information) instead of reporting a bad
  header and terminating.
* pg: Handle cases where standard output is a terminal but not open for
  reading (bug reported by Keith Doyle).
* cpio, tar: If an archive contains a file that is a hardlink to itself on
  extraction, do nothing at all instead of discarding its data. This mainly
  happens if a file with more than one link is included under the same name
  in an archive twice. Since special situations on extraction such as bind
  mounts on Linux may lead to this scenario even in cases where both link
  names are different, this is now treated like a successful link() step
  (bugreport by Thomas Graf).
* Made the getopt() substitute work with the getopt$UNIX2003() function
  redirection mechanism on Mac OS X Leopard.
* mknod, pgrep, ps: Mac OS X Leopard compilation fixes.
* w, whodo: On a 32-bit system where /proc/kcore is larger than 2 GB,
  avoid a harmless error message (bug reported by Pierre Gentile).
* ptime: Works on recent Linux kernels again.
* col: The behavior of the -x flag has been inverted to actually match
  the documentation, i.e. to suppress conversion of spaces to tabs (patch
  by Markus Schnalke).
* cut: With "-s", print the last field of a line that was also the last
  field specified with "-f" (bugreport by Pierre Gentile).

Release 240220
* [NEW] seq: Stone-simple implementation; needs some improvement yet.
* [NEW] watch: Complete, simple and stable implementation.
* [NEW] timeout: Complete implementation with extended payloads.
* readlink: Different help messages for both UCB and default variants.

Release 230503
* chroot: "docs: Update date at chroot.8 manual page."
* chroot: "chore: Remove comment, will move to GitHub"
* chroot: "docs: Rewrote chroot(8) manual page"
* chroot: "chore: Add ``usage()`` function at chroot.c"
* tabs: "fix: [t,g]roff formatting at tabs/fspec.5."
* readlink: "docs: Created new manual page for the non-UCB variant. (#24)"
* Makefile.mk: "fix: Don't symbolic/hard link readlink from UCBDIR to DEFDIR anymore"
* readlink: "docs: Add comments clarifying the C preprocessor directives."
* readlink: "fix: Ignore ``-n`` option when operating multiple files."
* readlink: "feat: Create non-UCB/GNU-alike variant."
* readlink: "fix: Made ``name_size`` private again, so ``pfmt()`` won't go segmentation fault."
* readlink: "feat: First step, start modularizing the code."
* libcommon: "fix: getdirentries replacement for OBSD >5.4"
* fix-me: For new releases, finish porting for OpenBSD and implement watch(1) and timeout(1).

Release 230305
* getconf: "Now _NPROCESSORS_ONLN and _NPROCESSORS_CONF are implemented."
* (*): "Removed the remaining licence headers, now every licence is referenced by its SPDX identifier."

Release 230228
* [NEW] readlink: New implementation, extensively based on OpenBSD's.
* readlink: It's installed at /usr/ucb per default and symbolic/hard linked to /usr/5bin.

Release 230120
* Notes for the library directories if linking against the GNU C library.

tabs/newform.c
/*
 * newform - change the format of a text file
 *
 * Gunnar Ritter, Freiburg i. Br., Germany, May 2003.
 */
/*
 * Copyright (c) 2003 Gunnar Ritter
 *
 * SPDX-Licence-Identifier: Zlib
 */

#if __GNUC__ >= 3 && __GNUC_MINOR__ >= 4 || __GNUC__ >= 4
#define	USED	__attribute__ ((used))
#elif defined __GNUC__
#define	USED	__attribute__ ((unused))
#else
#define	USED
#endif
static const char sccsid[] USED = "@(#)newform.sl	1.7 (gritter) 5/29/05";

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <libgen.h>
#include <locale.h>
#include <ctype.h>
#include <wchar.h>
#include <wctype.h>
#include <unistd.h>

#if defined (__GLIBC__) && defined (_IO_putc_unlocked)
#undef	putchar
#define	putchar(c)	_IO_putc_unlocked(c, stdout)
#endif

#include <iblok.h>
#include <mbtowi.h>
#include "tabspec.h"

#define	next(wc, s, n) (mb_cur_max > 1 && *(s) & 0200 ? \
			((n) = mbtowi(&(wc), (s), mb_cur_max), \
			 (n) = ((n) > 0 ? (n) : (n) < 0 ? (wc=WEOF, 1) : 1)) :\
		((wc) = *(s) & 0377, (n) = 1))

#define	check(lp)	((lp)->l_len < (lp)->l_size ? (lp)->l_line : \
		((lp)->l_line = srealloc((lp)->l_line, \
				((lp)->l_size += 128) * sizeof *(lp)->l_line)))

#define	add(lp, c)	(check(lp)[(lp)->l_len] = (c), (lp)->l_len++)

static struct 	line {
	wchar_t		*l_line;
	struct tabulator	*l_ltab;
	size_t		l_size;
	size_t		l_len;
	size_t		l_max;
} line1, line2;

static struct action {
	struct action	*a_nxt;
	struct line	*(*a_fcn)(struct line *, struct action *);
	struct tabulator	*a_tab;
	wchar_t	a_chr;
	long	a_val;
} *actions;

static const char	*progname;
static int		status;
static int		sflag;
static wint_t		cflag = ' ';
static int		mb_cur_max;

static int		scan(int, char **);
static void		addact(int, const char *);
static void		usage(void);
static void		newform(const char *);
static void		sproc(const char *, size_t);
static struct line	*iproc(struct line *, struct action *);
static struct line	*oproc(struct line *, struct action *);
static void		spaces(struct line *, long *, long *,
				struct tabulator **);
static struct line	*bproc(struct line *, struct action *);
static struct line	*eproc(struct line *, struct action *);
static struct line	*pproc(struct line *, struct action *);
static struct line	*aproc(struct line *, struct action *);
static struct line	*fproc(struct line *, struct action *);
static struct line	*lproc(struct line *, struct action *);
static void		print(const wchar_t *, size_t);
static struct line	*other(struct line *);
static void		toolarge(long);
static void		*srealloc(void *, size_t);

int
main(int argc, char **argv)
{
	int	i;

	progname = basename(argv[0]);
	setlocale(LC_CTYPE, "");
	mb_cur_max = MB_CUR_MAX;
	if (scan(argc, argv)) {
		for (i = 1; i < argc; i++)
			if (argv[i][0] != '-')
				newform(argv[i]);
	} else
		newform(NULL);
	return status;
}

static int
scan(int ac, char **av)
{
	int	i, filecnt = 0, n;

	for (i = 1; i < ac; i++)
		if (av[i][0] == '-')
			switch (av[i][1]) {
			case 's':
				sflag = 1;
				break;
			case 'c':
				if (av[i][2]) {
					next(cflag, &av[i][2], n);
					if (cflag == WEOF) {
						fprintf(stderr,
							"%s: illegal -cchar\n",
								progname);
						exit(1);
					}
				} else
					cflag = ' ';
				break;
			default:
				addact(av[i][1], &av[i][2]);
			}
		else
			filecnt++;
	return filecnt != 0;
}

static void
addact(int c, const char *s)
{
	struct action	*ap, *aq;

	ap = scalloc(1, sizeof *ap);
	switch (c) {
	case 'i':
	case 'o':
		ap->a_fcn = c == 'i' ? iproc : oproc;
		if (s[0] == '-' && s[1] == '0' && s[2] == '\0')
			ap->a_tab = NULL;
		else if ((ap->a_tab = tabstops(*s?s:"-8", 80)) == NULL) {
			switch (taberrno) {
			case TABERR_CANTOP:
				fprintf(stderr, "%s: can't open %s\n",
						progname, &s[2]);
				break;
			case TABERR_FILIND:
				fprintf(stderr, "%s: tabspec indirection "
						"illegal\n", progname);
				break;
			default:
				fprintf(stderr, "%s: tabspec in error\n"
			"tabspec is \t-a\t-a2\t-c\t-c2\t-c3\t-f\t-p\t-s\n"
			"\t\t-u\t--\t--file\t-number\tnumber,..,number\n",
					progname);
			}
			exit(1);
		}
		break;
	case 'b':
		ap->a_fcn = bproc;
		ap->a_val = atol(s);
		break;
	case 'e':
		ap->a_fcn = eproc;
		ap->a_val = atol(s);
		break;
	case 'p':
		ap->a_fcn = pproc;
		ap->a_val = atol(s);
		ap->a_chr = cflag;
		break;
	case 'a':
		ap->a_fcn = aproc;
		ap->a_val = atol(s);
		ap->a_chr = cflag;
		break;
	case 'f':
		ap->a_fcn = fproc;
		break;
	case 'l':
		ap->a_fcn = lproc;
		if ((ap->a_val = atol(s)) <= 0)
			ap->a_val = 72;
		break;
	default:
		usage();
	}
	if (actions) {
		for (aq = actions; aq->a_nxt; aq = aq->a_nxt);
		aq->a_nxt = ap;
	} else
		actions = ap;
}

static void
usage(void)
{
	fprintf(stderr, "\
usage: %s  [-s] [-itabspec] [-otabspec] [-pn] [-en] [-an] [-f] [-cchar]\n\
\t\t[-ln] [-bn] [file ...]\n",
		progname);
	exit(1);
}

static void
newform(const char *fn)
{
	struct iblok	*ip;
	static char	*line;
	static size_t	linesize;
	size_t	linelen;

	if ((ip = fn ? ib_open(fn, 0) : ib_alloc(0, 0)) == NULL) {
		fprintf(stderr, "%s: can't open %s\n", progname, fn);
		status |= 1;
		return;
	}
	while ((linelen = ib_getlin(ip, &line, &linesize, srealloc)) != 0) {
		if (line[linelen-1] == '\n')
			line[--linelen] = '\0';
		sproc(line, linelen);
	}
	if (ip->ib_fd)
		ib_close(ip);
	else
		ib_free(ip);
}

static void
sproc(const char *bline, size_t blen)
{
	struct line	*lp;
	struct action	*ap;
	const char	*bp;
	wchar_t	sheared[10];
	wint_t	c = WEOF;
	long	i;
	int	n;

	bp = bline;
	if (sflag) {
		for (i = 1; i <= 9 && bp < &bline[blen]; i++) {
			next(c, bp, n);
			bp += n;
			if (c == '\t') {
				i--;
				break;
			}
			if (c == WEOF)
				i--;
			else
				sheared[i] = c;
		}
		if (i == 10) {
			sheared[i=8] = '*';
			while (bp < &bline[blen]) {
				next(c, bp, n);
				bp += n;
				if (c == '\t')
					break;
			}
		}
		if (c != '\t') {
			fprintf(stderr, "not -s format\n");
			exit(1);
		}
		sheared[0] = i;
	}
	lp = &line1;
	lp->l_len = 0;
	while (bp < &bline[blen]) {
		next(c, bp, n);
		bp += n;
		if (c == WEOF)
			continue;
		add(lp, c);
	}
	lp->l_max = 80;
	lp->l_ltab = 0;
	for (ap = actions; ap; ap = ap->a_nxt)
		lp = ap->a_fcn(lp, ap);
	print(lp->l_line, lp->l_len);
	if (sflag && sheared[0])
		print(&sheared[1], sheared[0]);
	putchar('\n');
}

static struct line *
iproc(struct line *lp, struct action *ap)
{
	struct line	*np = other(lp);
	struct tabulator	*tp = ap->a_tab;
	wchar_t	c;
	long	i, col;
	int	w;

	col = 0;
	for (i = 0; i < lp->l_len; i++) {
		switch (c = lp->l_line[i]) {
		case '\b':
			if (col > 0)
				col--;
			add(np, c);
			break;
		case '\t':
			if (ap->a_tab) {
				if (tp && tp->t_rep) {
					if (col % tp->t_rep == 0) {
						add(np, ' ');
						col++;
					}
					while (col % tp->t_rep) {
						add(np, ' ');
						col++;
					}
					break;
				}
				while (tp && (col>tp->t_tab || tp->t_tab==0))
					tp = tp->t_nxt;
				if (tp && col == tp->t_tab) {
					add(np, ' ');
					col++;
					tp = tp->t_nxt;
				}
				if (tp) {
					while (col < tp->t_tab) {
						add(np, ' ');
						col++;
					}
					tp = tp->t_nxt;
				}
				break;
			} else if (ap->a_tab)
				break;
			c = ' ';
			/*FALLTHRU*/
		default:
			if (mb_cur_max > 1 && (w = wcwidth(c)) >= 0)
				col += w;
			else
				col++;
			add(np, c);
			break;
		}
	}
	return np;
}

static struct line *
oproc(struct line *lp, struct action *ap)
{
	struct line	*np = other(lp);
	struct tabulator	*tp = ap->a_tab;
	wchar_t	c;
	long	i, col, spc;
	int	w;

	np->l_ltab = ap->a_tab;
	col = 0, spc = 0;
	for (i = 0; i < lp->l_len; i++) {
		c = lp->l_line[i];
		if (c == ' ' && tp) {
			spc++;
			continue;
		}
		if (spc)
			spaces(np, &spc, &col, &tp);
		switch (c) {
		case '\b':
			if (col > 0)
				col--;
			break;
		case '\t':
			if (ap->a_tab) {
				if (tp && tp->t_rep) {
					if (col % tp->t_rep == 0)
						col++;
					while (col % tp->t_rep)
						col++;
					break;
				}
				while (tp && (col>tp->t_tab || tp->t_tab==0))
					tp = tp->t_nxt;
				if (tp && col == tp->t_tab) {
					col++;
					tp = tp->t_nxt;
				}
				if (tp) {
					while (col < tp->t_tab)
						col++;
					tp = tp->t_nxt;
				}
			} else if (ap->a_tab == 0) {
				c = ' ';
				col++;
			}
			break;
		default:
			if (mb_cur_max > 1 && (w = wcwidth(c)) >= 0)
				col += w;
			else
				col++;
		}
		add(np, c);
	}
	if (spc)
		spaces(np, &spc, &col, &tp);
	return np;
}

static void
spaces(struct line *np, long *sp, long *cp, struct tabulator **tp)
{
	long	nsp;

	if (*tp && (*tp)->t_rep == 0) {
		while (*tp && (*tp)->t_tab <= *cp)
			*tp = (*tp)->t_nxt;
		while (*sp && *tp && (*cp + *sp) >= (*tp)->t_tab) {
			add(np, '\t');
			*sp -= (*tp)->t_tab - *cp;
			*cp = (*tp)->t_tab;
			*tp = (*tp)->t_nxt;
		}
	} else if (*tp) {
		while (*sp >= (nsp = (*tp)->t_rep - *cp % (*tp)->t_rep)) {
			add(np, '\t');
			*sp -= nsp;
			*cp += nsp;
		}
	}
	while (*sp) {
		add(np, ' ');
		(*cp)++;
		(*sp)--;
	}
}

static struct line *
bproc(struct line *lp, struct action *ap)
{
	struct line	*np;
	long	i, min;

	if (lp->l_len > lp->l_max) {
		min = ap->a_val>0 ? ap->a_val : lp->l_len-lp->l_max;
		if (min >= lp->l_len) {
			toolarge(lp->l_len);
			return lp;
		}
		np = other(lp);
		for (i = min; i < lp->l_len; i++)
			add(np, lp->l_line[i]);
		return np;
	} else
		return lp;
}

static struct line *
eproc(struct line *lp, struct action *ap)
{
	if (lp->l_len > lp->l_max) {
		if (ap->a_val > 0) {
			if (ap->a_val > lp->l_len)
				toolarge(lp->l_len);
			else
				lp->l_len -= ap->a_val;
		} else
			lp->l_len = lp->l_max;
	}
	return lp;
}

static struct line *
pproc(struct line *lp, struct action *ap)
{
	struct line	*np;
	long	i;

	if (lp->l_len < lp->l_max) {
		np = other(lp);
		i = ap->a_val>0 ? ap->a_val : lp->l_max - lp->l_len;
		while (i--)
			add(np, ap->a_chr);
		for (i = 0; i < lp->l_len; i++)
			add(np, lp->l_line[i]);
		return np;
	} else
		return lp;
}

static struct line *
aproc(struct line *lp, struct action *ap)
{
	long	i;

	if (lp->l_len < lp->l_max) {
		i = ap->a_val>0 ? ap->a_val : lp->l_max - lp->l_len;
		while (i--)
			add(lp, ap->a_chr);
	}
	return lp;
}

static struct line *
fproc(struct line *lp, struct action *ap)
{
	if (ap->a_val == 0) {
		printf("<:t%s d:>\n",
			lp->l_ltab&&lp->l_ltab->t_str?lp->l_ltab->t_str:"-8");
		ap->a_val = 1;
	}
	return lp;
}

static struct line *
lproc(struct line *lp, struct action *ap)
{
	lp->l_max = ap->a_val ? ap->a_val : 72;
	return lp;
}

static void
print(const wchar_t *line, size_t length)
{
	const wchar_t	*sp;
	char	mb[MB_LEN_MAX];
	int	i, n;

	for (sp = line; sp < &line[length]; sp++) {
		if (mb_cur_max > 1 && *sp & ~(wchar_t)0177) {
			n = wctomb(mb, *sp);
			for (i = 0; i < n; i++)
				putchar(mb[i] & 0377);
		} else
			putchar(*sp & 0377);
	}
}

static struct line *
other(struct line *lp)
{
	struct line *np;

	np = lp == &line1 ? &line2 : &line1;
	np->l_ltab = lp->l_ltab;
	np->l_len = 0;
	np->l_max = lp->l_max;
	return np;
}

static void
toolarge(long n)
{
	fprintf(stderr, "%s: truncate request larger than line, %ld\n",
			progname, n);
}

static void *
srealloc(void *op, size_t size)
{
	void	*np;

	if ((np = realloc(op, size)) == NULL) {
		write(2, "no memory\n", 10);
		_exit(077);
	}
	return np;
}
